<a href="https://colab.research.google.com/github/LokeRuiKee/ml-dl-playground/blob/main/CNN_experiment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CNN Template Code

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

In [ ]:
# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


11490434/11490434 [==============================] - 0s 0us/step
Epoch 1/5
1875/1875 [==============================] - 58s 30ms/step - loss: 0.2013 - accuracy: 0.9384
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0778 - accuracy: 0.9767
Epoch 3/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0586 - accuracy: 0.9826
Epoch 4/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0453 - accuracy: 0.9860
Epoch 5/5
313/313 [==============================] - 5s 14ms/step - loss: 0.0255 - accuracy: 0.9914
Test accuracy: 0.9914000034332275


# 7 activation, optimizer = adam

## relu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 58s 30ms/step - loss: 0.2063 - accuracy: 0.9369
Epoch 2/5
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0790 - accuracy: 0.9766
Epoch 3/5
1875/1875 [==============================] - 57s 31ms/step - loss: 0.0576 - accuracy: 0.9825
Epoch 4/5
1875/1875 [==============================] - 55s 29ms/step - loss: 0.0466 - accuracy: 0.9861
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 0.0236 - accuracy: 0.9928
Test accuracy: 0.9927999973297119


## LeakyReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='LeakyReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='LeakyReLU'),                          # Convolutional layer with 64 filters of size 3x3 and LeakyReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='LeakyReLU'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 61s 32ms/step - loss: 0.1544 - accuracy: 0.9530
Epoch 2/5
1875/1875 [==============================] - 63s 33ms/step - loss: 0.0628 - accuracy: 0.9808
Epoch 3/5
1875/1875 [==============================] - 61s 32ms/step - loss: 0.0513 - accuracy: 0.9843
Epoch 4/5
1875/1875 [==============================] - 64s 34ms/step - loss: 0.0402 - accuracy: 0.9870
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0325 - accuracy: 0.9894
Test accuracy: 0.9894000291824341


## elu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='elu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and elu activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='elu'),                          # Convolutional layer with 64 filters of size 3x3 and elu activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='elu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 68s 36ms/step - loss: 0.1607 - accuracy: 0.9517
Epoch 2/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.0736 - accuracy: 0.9777
Epoch 3/5
1875/1875 [==============================] - 58s 31ms/step - loss: 0.0594 - accuracy: 0.9815
Epoch 4/5
1875/1875 [==============================] - 59s 32ms/step - loss: 0.0508 - accuracy: 0.9850
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 0.0324 - accuracy: 0.9910
Test accuracy: 0.9909999966621399


## PReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='PReLU', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and PReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='PReLU'),                          # Convolutional layer with 64 filters of size 3x3 and PReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='PReLU'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 124s 36ms/step - loss: 0.1842 - accuracy: 0.9438
Epoch 2/5
1875/1875 [==============================] - 71s 38ms/step - loss: 0.0630 - accuracy: 0.9806
Epoch 3/5
1875/1875 [==============================] - 78s 41ms/step - loss: 0.0475 - accuracy: 0.9853
Epoch 4/5
1875/1875 [==============================] - 81s 43ms/step - loss: 0.0382 - accuracy: 0.9883
Epoch 5/5
313/313 [==============================] - 4s 11ms/step - loss: 0.0298 - accuracy: 0.9904
Test accuracy: 0.9904000163078308


## Sigmoid

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='sigmoid', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and sigmoid activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='sigmoid'),                          # Convolutional layer with 64 filters of size 3x3 and sigmoid activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='sigmoid'),                                   # Fully connected layer with 128 neurons and sigmoid activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 57s 30ms/step - loss: 1.2158 - accuracy: 0.5532
Epoch 2/5
1875/1875 [==============================] - 57s 30ms/step - loss: 0.1618 - accuracy: 0.9529
Epoch 3/5
1875/1875 [==============================] - 55s 29ms/step - loss: 0.1130 - accuracy: 0.9664
Epoch 4/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0909 - accuracy: 0.9726
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 0.0525 - accuracy: 0.9819
Test accuracy: 0.9818999767303467


## tanh

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='tanh', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and tanh activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='tanh'),                          # Convolutional layer with 64 filters of size 3x3 and tanh activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='tanh'),                                   # Fully connected layer with 128 neurons and tanh activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.1661 - accuracy: 0.9503
Epoch 2/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0747 - accuracy: 0.9776
Epoch 3/5
1875/1875 [==============================] - 62s 33ms/step - loss: 0.0622 - accuracy: 0.9813
Epoch 4/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0552 - accuracy: 0.9833
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0397 - accuracy: 0.9862
Test accuracy: 0.9861999750137329


## softmax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='softmax', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and softmax activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='softmax'),                          # Convolutional layer with 64 filters of size 3x3 and softmax activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='softmax'),                                   # Fully connected layer with 128 neurons and softmax activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 73s 39ms/step - loss: 1.9056 - accuracy: 0.2966
Epoch 2/5
1875/1875 [==============================] - 71s 38ms/step - loss: 1.4360 - accuracy: 0.4396
Epoch 3/5
1875/1875 [==============================] - 63s 34ms/step - loss: 1.2461 - accuracy: 0.5382
Epoch 4/5
1875/1875 [==============================] - 62s 33ms/step - loss: 1.1331 - accuracy: 0.5700
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 0.3882 - accuracy: 0.9667
Test accuracy: 0.96670001745224


## conclusion
Best activation function: relu. Test accuracy: 0.9927999973297119

# Best activation function: relu, 8 optimizer

## SGD

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='SGD',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 53s 28ms/step - loss: 0.6601 - accuracy: 0.7931
Epoch 2/5
1875/1875 [==============================] - 52s 28ms/step - loss: 0.2274 - accuracy: 0.9311
Epoch 3/5
1875/1875 [==============================] - 55s 29ms/step - loss: 0.1675 - accuracy: 0.9498
Epoch 4/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.1383 - accuracy: 0.9582
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 0.0621 - accuracy: 0.9791
Test accuracy: 0.9790999889373779


## SGD with Momentum

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
SGDm = tf.keras.optimizers.SGD(learning_rate=0.01, momentum=0.9)
model.compile(optimizer= SGDm,
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 53s 28ms/step - loss: 0.2684 - accuracy: 0.9147
Epoch 2/5
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0883 - accuracy: 0.9730
Epoch 3/5
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0681 - accuracy: 0.9797
Epoch 4/5
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0552 - accuracy: 0.9834
Epoch 5/5
313/313 [==============================] - 3s 8ms/step - loss: 0.0291 - accuracy: 0.9897
Test accuracy: 0.9897000193595886


## RMSProp

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='RMSProp',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 52s 27ms/step - loss: 0.1838 - accuracy: 0.9439
Epoch 2/5
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0680 - accuracy: 0.9803
Epoch 3/5
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0529 - accuracy: 0.9843
Epoch 4/5
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0476 - accuracy: 0.9860
Epoch 5/5
313/313 [==============================] - 5s 15ms/step - loss: 0.0342 - accuracy: 0.9895
Test accuracy: 0.9894999861717224


## adam

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 52s 27ms/step - loss: 0.2087 - accuracy: 0.9363
Epoch 2/5
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0765 - accuracy: 0.9776
Epoch 3/5
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0557 - accuracy: 0.9833
Epoch 4/5
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0477 - accuracy: 0.9858
Epoch 5/5
313/313 [==============================] - 3s 8ms/step - loss: 0.0228 - accuracy: 0.9925
Test accuracy: 0.9925000071525574


## adamax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adamax',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 52s 27ms/step - loss: 0.3472 - accuracy: 0.8949
Epoch 2/5
1875/1875 [==============================] - 52s 28ms/step - loss: 0.1291 - accuracy: 0.9615
Epoch 3/5
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0984 - accuracy: 0.9713
Epoch 4/5
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0799 - accuracy: 0.9764
Epoch 5/5
313/313 [==============================] - 3s 8ms/step - loss: 0.0327 - accuracy: 0.9885
Test accuracy: 0.9884999990463257


## nadam

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='nadam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 54s 28ms/step - loss: 0.1943 - accuracy: 0.9421
Epoch 2/5
1875/1875 [==============================] - 54s 29ms/step - loss: 0.0737 - accuracy: 0.9782
Epoch 3/5
1875/1875 [==============================] - 54s 29ms/step - loss: 0.0533 - accuracy: 0.9846
Epoch 4/5
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0449 - accuracy: 0.9865
Epoch 5/5
313/313 [==============================] - 3s 8ms/step - loss: 0.0265 - accuracy: 0.9904
Test accuracy: 0.9904000163078308


## AdaGrad

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='AdaGrad',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 51s 27ms/step - loss: 1.5277 - accuracy: 0.5276
Epoch 2/5
1875/1875 [==============================] - 50s 27ms/step - loss: 0.6179 - accuracy: 0.8183
Epoch 3/5
1875/1875 [==============================] - 53s 28ms/step - loss: 0.4564 - accuracy: 0.8647
Epoch 4/5
1875/1875 [==============================] - 50s 27ms/step - loss: 0.3802 - accuracy: 0.8874
Epoch 5/5
313/313 [==============================] - 3s 8ms/step - loss: 0.2078 - accuracy: 0.9389
Test accuracy: 0.9388999938964844


## AdaDelta

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='AdaDelta',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 53s 28ms/step - loss: 2.2759 - accuracy: 0.1522
Epoch 2/5
1875/1875 [==============================] - 53s 28ms/step - loss: 2.2155 - accuracy: 0.2628
Epoch 3/5
1875/1875 [==============================] - 55s 29ms/step - loss: 2.1475 - accuracy: 0.3632
Epoch 4/5
1875/1875 [==============================] - 53s 28ms/step - loss: 2.0584 - accuracy: 0.4415
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 1.8504 - accuracy: 0.7042
Test accuracy: 0.704200029373169


## conclusion
Best optimizer: Adam. Test accuracy: 0.9925000071525574

#Experimenting Best Output Activation Function


## output relu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='relu')                                  # Output layer with 10 neurons for classification (relu activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 64s 34ms/step - loss: 2.4469 - accuracy: 0.1229
Epoch 2/5
1875/1875 [==============================] - 54s 29ms/step - loss: 2.3091 - accuracy: 0.0987
Epoch 3/5
1875/1875 [==============================] - 52s 28ms/step - loss: 2.3049 - accuracy: 0.0987
Epoch 4/5
1875/1875 [==============================] - 51s 27ms/step - loss: 2.3037 - accuracy: 0.0987
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 2.3026 - accuracy: 0.0980
Test accuracy: 0.09799999743700027


## output LeakyReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='LeakyReLU')                                  # Output layer with 10 neurons for classification (LeakyReLU activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 62s 32ms/step - loss: 2.4100 - accuracy: 0.1373
Epoch 2/5
1875/1875 [==============================] - 59s 31ms/step - loss: 2.3062 - accuracy: 0.1087
Epoch 3/5
1875/1875 [==============================] - 52s 28ms/step - loss: 2.3037 - accuracy: 0.1090
Epoch 4/5
1875/1875 [==============================] - 51s 27ms/step - loss: 2.3035 - accuracy: 0.1086
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 2.3026 - accuracy: 0.1009
Test accuracy: 0.10090000182390213


## output elu

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='elu')                                  # Output layer with 10 neurons for classification (elu activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 52s 27ms/step - loss: 2.4055 - accuracy: 0.1079
Epoch 2/5
1875/1875 [==============================] - 51s 27ms/step - loss: 2.3027 - accuracy: 0.1037
Epoch 3/5
1875/1875 [==============================] - 52s 28ms/step - loss: 2.3026 - accuracy: 0.1046
Epoch 4/5
1875/1875 [==============================] - 51s 27ms/step - loss: 2.3026 - accuracy: 0.1060
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 2.3026 - accuracy: 0.1032
Test accuracy: 0.10320000350475311


## output PReLU

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='PReLU')                                  # Output layer with 10 neurons for classification (PReLU activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 54s 28ms/step - loss: 2.3867 - accuracy: 0.0991
Epoch 2/5
1875/1875 [==============================] - 51s 27ms/step - loss: 2.3032 - accuracy: 0.0987
Epoch 3/5
1875/1875 [==============================] - 51s 27ms/step - loss: 2.3032 - accuracy: 0.0987
Epoch 4/5
1875/1875 [==============================] - 50s 27ms/step - loss: 2.3040 - accuracy: 0.0987
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 2.3026 - accuracy: 0.0980
Test accuracy: 0.09799999743700027


## output sigmoid

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='sigmoid')                                  # Output layer with 10 neurons for classification (sigmoid activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 53s 28ms/step - loss: 0.2012 - accuracy: 0.9390
Epoch 2/5
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0734 - accuracy: 0.9773
Epoch 3/5
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0569 - accuracy: 0.9833
Epoch 4/5
1875/1875 [==============================] - 53s 29ms/step - loss: 0.0446 - accuracy: 0.9865
Epoch 5/5
313/313 [==============================] - 4s 13ms/step - loss: 0.0256 - accuracy: 0.9927
Test accuracy: 0.9926999807357788


## output tanh

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='tanh')                                  # Output layer with 10 neurons for classification (tanh activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 58s 31ms/step - loss: 2.4749 - accuracy: 0.1280
Epoch 2/5
1875/1875 [==============================] - 57s 30ms/step - loss: 2.3082 - accuracy: 0.1228
Epoch 3/5
1875/1875 [==============================] - 59s 32ms/step - loss: 2.3049 - accuracy: 0.1177
Epoch 4/5
1875/1875 [==============================] - 58s 31ms/step - loss: 2.3041 - accuracy: 0.1098
Epoch 5/5
313/313 [==============================] - 3s 9ms/step - loss: 2.3026 - accuracy: 0.0891
Test accuracy: 0.08910000324249268


## output softmax

In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
# Define the architecture of the CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),  # Convolutional layer with 32 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Conv2D(64, (3, 3), activation='relu'),                          # Convolutional layer with 64 filters of size 3x3 and ReLU activation
    MaxPooling2D((2, 2)),                                           # Max pooling layer with pool size 2x2
    Flatten(),                                                       # Flatten layer to convert 2D feature maps to 1D
    Dense(128, activation='relu'),                                   # Fully connected layer with 128 neurons and ReLU activation
    Dropout(0.5),                                                    # Dropout layer with dropout rate 0.5 for regularization
    Dense(10, activation='softmax')                                  # Output layer with 10 neurons for classification (softmax activation)
])
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',   # Using cross-entropy loss for classification
              metrics=['accuracy'])

# Load and preprocess the data (example: MNIST dataset)
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()
# Reshape and normalize the data
x_train = x_train.reshape(-1, 28, 28, 1) / 255.0
x_test = x_test.reshape(-1, 28, 28, 1) / 255.0
# Train the model
model.fit(x_train, y_train, epochs=5)
# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print('Test accuracy:', test_acc)


Epoch 1/5
1875/1875 [==============================] - 60s 32ms/step - loss: 0.2028 - accuracy: 0.9380
Epoch 2/5
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0748 - accuracy: 0.9779
Epoch 3/5
1875/1875 [==============================] - 61s 33ms/step - loss: 0.0534 - accuracy: 0.9845
Epoch 4/5
1875/1875 [==============================] - 57s 31ms/step - loss: 0.0428 - accuracy: 0.9874
Epoch 5/5
313/313 [==============================] - 3s 10ms/step - loss: 0.0268 - accuracy: 0.9925
Test accuracy: 0.9925000071525574


## conclusion
Best output activation function: sigmoid. Test accuracy: 0.9926999807357788